src/Solver/time_stepping/GhostTimeClusterWithCopy.cpp

```diff
@@ -0,0 +1,199 @@
+    }
+  }
+}
+
+template <MPI::DataTransferMode CommType>
+void GhostTimeClusterWithCopy<CommType>::sendCopyLayer() {
+  SCOREP_USER_REGION("sendCopyLayer", SCOREP_USER_REGION_TYPE_FUNCTION)
+  assert(ct.correctionTime > lastSendTime);
+
+  lastSendTime = ct.correctionTime;
+  auto prefetchedRegions = prefetchCopyLayer();
+
+  while (not prefetchedRegions.empty()) {
+    for (auto region = prefetchedRegions.begin(); region != prefetchedRegions.end();) {
+      auto* stream = prefetchCopyRegionsStreams[*region];
+      if (device.api->isStreamWorkDone(stream)) {
+        MPI_Isend(duplicatedCopyRegions[*region],
+                  static_cast<int>(meshStructure->copyRegionSizes[*region]),
+                  MPI_C_REAL,
+                  meshStructure->neighboringClusters[*region][0],
+                  timeData + meshStructure->sendIdentifiers[*region],
+                  seissol::MPI::mpi.comm(),
+                  meshStructure->sendRequests + (*region));
+        sendQueue.push_back(*region);
+        region = prefetchedRegions.erase(region);
+      } else {
+        ++region;
+      }
+    }
+  }
+}
+
+template <MPI::DataTransferMode CommType>
+void GhostTimeClusterWithCopy<CommType>::receiveGhostLayer() {
+  SCOREP_USER_REGION("receiveGhostLayer", SCOREP_USER_REGION_TYPE_FUNCTION)
+  assert(ct.predictionTime > lastSendTime);
+  for (unsigned int region = 0; region < numberOfRegions; ++region) {
+    if (meshStructure->neighboringClusters[region][1] == static_cast<int>(otherGlobalClusterId)) {
+      MPI_Irecv(duplicatedGhostRegions[region],
+                static_cast<int>(meshStructure->ghostRegionSizes[region]),
+                MPI_C_REAL,
+                meshStructure->neighboringClusters[region][0],
+                timeData + meshStructure->receiveIdentifiers[region],
+                seissol::MPI::mpi.comm(),
+                meshStructure->receiveRequests + region);
+      receiveRegionsStates[region] = RequiresMpiTesting;
+      receiveQueue.push_back(region);
+    }
+  }
+}
+
+template <MPI::DataTransferMode CommType>
+bool GhostTimeClusterWithCopy<CommType>::testReceiveQueue() {
+  for (auto region = receiveQueue.begin(); region != receiveQueue.end();) {
+    const auto state = receiveRegionsStates[*region];
+
+    switch (state) {
+    case ReceiveState::RequiresMpiTesting: {
+      int testSuccess = 0;
+      MPI_Request* request = &(meshStructure->receiveRequests)[*region];
+      MPI_Test(request, &testSuccess, MPI_STATUS_IGNORE);
+      if (testSuccess) {
+        prefetchGhostRegion(*region);
+        receiveRegionsStates[*region] = RequiresPrefetchTesting;
+      }
+      ++region;
+      break;
+    }
+    case ReceiveState::RequiresPrefetchTesting: {
+      auto* stream = prefetchGhostRegionsStreams[*region];
+      if (device.api->isStreamWorkDone(stream)) {
+        receiveRegionsStates[*region] = Ready;
+        // Note: fall to `Ready` is intentional
```

Comment:
I don't get this comment

Reply:
Ah, you mean fallthrough to ready is intentional? Yeah, makes sense.

Reply:
yes, `[[fallthrough]]` only works inside `switch`-statements. I just edited the comment